File: dashboards/django/views/dlp.py
"""
DLP view: manage data loss prevention rules, violations, and scanning.
"""

from django.http import HttpResponse, HttpResponseRedirect
from django.template.loader import render_to_string

from utils.api import api_request
from utils.helpers import badge, page_context


def dlp_view(request):
    """Handle DLP rules, violations, and scan actions."""
    if not request.session.get('token'):
        return HttpResponseRedirect('/login')

    token = request.session['token']

    if request.method == 'POST':
        action = request.POST.get('_action')
        if action == 'create_rule':
            body = {
                'name': request.POST['name'],
                'pattern': request.POST['pattern'],
                'severity': request.POST.get('severity', 'medium'),
            }
            result = api_request('POST', '/engine/dlp/rules', body=body, token=token)
            request.session['flash'] = 'Rule created!' if 'id' in result else result.get('error', 'Failed')
        elif action == 'delete_rule':
            rule_id = request.POST.get('rule_id')
            api_request('DELETE', f'/engine/dlp/rules/{rule_id}', token=token)
            request.session['flash'] = 'Rule deleted'
        elif action == 'scan':
            body = {'content': request.POST.get('content', '')}
            result = api_request('POST', '/engine/dlp/scan', body=body, token=token)
            request.session['flash'] = f"Scan complete: {result.get('matches', 0)} matches found" if 'matches' in result else result.get('error', 'Scan failed')
        return HttpResponseRedirect('/dlp')

    rules_data = api_request('GET', '/engine/dlp/rules?orgId=default', token=token)
    rules = rules_data.get('rules', [])

    violations_data = api_request('GET', '/engine/dlp/violations?orgId=default', token=token)
    violations = violations_data.get('violations', [])

    for r in rules:
        r['severity_badge'] = badge(r.get('severity', 'medium'))

    ctx = page_context(request, 'Data Loss Prevention', 'dlp')
    ctx['rules'] = rules
    ctx['violations'] = violations

    html = render_to_string('dlp.html', ctx)
    return HttpResponse(html)

File: dashboards/rails/routes/dlp.rb
# frozen_string_literal: true

# DLP Controller

get '/dlp' do
  require_auth!
  rules_res = api_get('/engine/dlp/rules')
  violations_res = api_get('/engine/dlp/violations')
  @rules = rules_res[:status] == 200 ? (rules_res[:body]['rules'] || rules_res[:body]) : []
  @rules = @rules.is_a?(Array) ? @rules : []
  @violations = violations_res[:status] == 200 ? (violations_res[:body]['violations'] || violations_res[:body]) : []
  @violations = @violations.is_a?(Array) ? @violations : []
  erb :dlp
end

post '/dlp/rules' do
  require_auth!
  res = api_post('/engine/dlp/rules', { name: params[:name], pattern: params[:pattern], action: params[:action], severity: params[:severity] })
  set_flash(res[:status] < 300 ? 'DLP rule created' : (res[:body]['error'] || 'Failed'), res[:status] < 300 ? 'success' : 'danger')
  redirect '/dlp'
end

post '/dlp/rules/:id/delete' do
  require_auth!
  res = api_delete("/engine/dlp/rules/#{params[:id]}")
  set_flash(res[:status] < 300 ? 'DLP rule deleted' : (res[:body]['error'] || 'Failed'), res[:status] < 300 ? 'success' : 'danger')
  redirect '/dlp'
end

post '/dlp/scan' do
  require_auth!
  res = api_post('/engine/dlp/scan', { content: params[:content] })
  if res[:status] < 300
    matches = res[:body]['matches'] || res[:body]['violations'] || []
    set_flash(matches.empty? ? 'Scan clean — no violations detected' : "Scan found #{matches.size} violation(s)", matches.empty? ? 'success' : 'warning')
  else
    set_flash(res[:body]['error'] || 'Scan failed', 'danger')
  end
  redirect '/dlp'
end
